### PR TITLE
Fixed race condition in OnEnd and added a unit test

### PR DESCRIPTION
**Pull request by Kaushal28**
Fixes: #3948

Fixed race condition in `OnEnd` and added a unit test.

### CODE REVIEW

File: sdk/trace/simple_span_processor_test.go [hunk 1]
```diff
@@ -150,6 +151,36 @@ func TestSimpleSpanProcessorShutdownOnEndConcurrency(t *testing.T) {
 	<-done
 }
 
+func TestSimpleSpanProcessorShutdownOnEndRace(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (pellared) [reply sdk/trace/simple_span_processor_test.go:154]: This is not a proper test. It only shows that `testExporter` is not concurrent safe. The issue is with the Jaeger exporter.

You can consider making `testExporter` concurrent safe (which is required by the `SpanExpoter` interface), but it is not related to the issue you are trying to solve.

**Review by pellared — DISMISSED** [review]

File: sdk/trace/simple_span_processor_test.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (Kaushal28) [reply sdk/trace/simple_span_processor_test.go:154]: Thanks for the feedback. So if it's not related, what do you suggest for testing this specific change?

[the remark this answers is not in the document]
Comment (pellared) [reply sdk/trace/simple_span_processor_test.go:154]: Maybe I am wrong and fix and tests are OK.  Personally, I need clarification. I created https://github.com/open-telemetry/opentelemetry-go/issues/3950

File: sdk/trace/simple_span_processor.go [hunk 2]
```diff
@@ -54,8 +54,8 @@ func (ssp *simpleSpanProcessor) OnStart(context.Context, ReadWriteSpan) {}
 
 // OnEnd immediately exports a ReadOnlySpan.
 func (ssp *simpleSpanProcessor) OnEnd(s ReadOnlySpan) {
-	ssp.exporterMu.RLock()
-	defer ssp.exporterMu.RUnlock()
+	ssp.exporterMu.Lock()
+	defer ssp.exporterMu.Unlock()
```
Comment (dmathieu) [inline sdk/trace/simple_span_processor.go:58]: This is the only use of `RLock` and `RUnlock`. So if we keep this approach, the `RWMutex` can be changed to a `Mutex`.

I'm not sure this is the proper approach though. The jaeger exporter should be able to run concurrently (maybe it's missing a mutex somewhere?), ensuring it shouldn't be the span processor's responsibility.

Comment (pellared) [reply sdk/trace/simple_span_processor.go:58]: PTAL https://github.com/open-telemetry/opentelemetry-go/issues/3950

File: sdk/trace/simple_span_processor_test.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (pellared) [reply sdk/trace/simple_span_processor_test.go:154]: @Kaushal28 We discussed the PR during the SIG meeting. The test and fix is correct. I have closed https://github.com/open-telemetry/opentelemetry-go/issues/3950

File: sdk/trace/simple_span_processor_test.go [hunk 3]
```diff
@@ -150,6 +151,36 @@ func TestSimpleSpanProcessorShutdownOnEndConcurrency(t *testing.T) {
...
+	go func() {
+		_, span := tp.Tracer("test").Start(context.Background(), "A")
+		defer span.End()
+		wg.Done()
+	}()
+
+	wg.Wait()
+
+	if err := ssp.Shutdown(context.Background()); err != nil {
+		t.Errorf("shutting the SimpleSpanProcessor down: %v", err)
```
Comment (pellared) [inline sdk/trace/simple_span_processor_test.go:177]: We are using testify in this repository. Please use `assert.NoError` instead,

File: sdk/trace/simple_span_processor_test.go [hunk 4]
```diff
@@ -150,6 +151,36 @@ func TestSimpleSpanProcessorShutdownOnEndConcurrency(t *testing.T) {
...
+	}
+	if !exporter.shutdown {
```
Comment (pellared) [inline sdk/trace/simple_span_processor_test.go:179]: We are using testify in this repository. Please use `assert.True` instead

File: sdk/trace/simple_span_processor_test.go [hunk 5]
```diff
@@ -150,6 +151,36 @@ func TestSimpleSpanProcessorShutdownOnEndConcurrency(t *testing.T) {
...
+	ssp := sdktrace.NewSimpleSpanProcessor(exporter)
+	tp := basicTracerProvider(t)
+	tp.RegisterSpanProcessor(ssp)
+
+	var wg sync.WaitGroup
+	wg.Add(2)
+	go func() {
+		_, span := tp.Tracer("test").Start(context.Background(), "A")
+		defer span.End()
+		wg.Done()
```
Comment (pellared) [inline sdk/trace/simple_span_processor_test.go:165]: I would rather add `defer wg.Done()` in the beginning of the function (safer if the goroutine would panic)

**Review by pellared — COMMENTED** [review]
Thanks for the PR 👍 

Can you also update the `CHANGELOG.md`?

File: sdk/trace/simple_span_processor.go [hunk 6]
```diff
@@ -54,8 +54,8 @@ func (ssp *simpleSpanProcessor) OnStart(context.Context, ReadWriteSpan) {}
 
 // OnEnd immediately exports a ReadOnlySpan.
 func (ssp *simpleSpanProcessor) OnEnd(s ReadOnlySpan) {
-	ssp.exporterMu.RLock()
-	defer ssp.exporterMu.RUnlock()
+	ssp.exporterMu.Lock()
```
[the remark this answers is not in the document]
Comment (Kaushal28) [reply sdk/trace/simple_span_processor.go:57]: Changed!

**Comment by Kaushal28** [comment]
Hey @pellared , I have made all the changes as per your comments and also updated `CHANGELOG.md`. Please have a look. Thanks!

File: CHANGELOG.md [hunk 7]
```diff
@@ -23,6 +23,7 @@ This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.htm
 - `TracerProvider` allows calling `Tracer()` while it's shutting down.
   It used to deadlock. (#3924)
 - Use the SDK version for the Telemetry SDK resource detector in `go.opentelemetry.io/otel/sdk/resource`. (#3949)
+- Data race in `exporters/jaeger` when calling `End()` on spans from concurrent goroutines. (#3951)
```
Comment (pellared) [inline CHANGELOG.md:26]:
```suggestion
- Fix a data race in `SpanProcessor` returned by `NewSimpleSpanProcessor` in `go.opentelemetry.io/otel/sdk/trace`. (#3951)
```

**Review by pellared — APPROVED** [review]
Thanks for your contribution 👍

File: sdk/trace/simple_span_processor_test.go [hunk 8]
```diff
@@ -150,6 +153,30 @@ func TestSimpleSpanProcessorShutdownOnEndConcurrency(t *testing.T) {
...
+	tp := basicTracerProvider(t)
+	tp.RegisterSpanProcessor(ssp)
+
+	var wg sync.WaitGroup
+	wg.Add(2)
+
+	span := func(spanName string) {
+		defer wg.Done()
+		_, span := tp.Tracer("test").Start(context.Background(), spanName)
+		defer span.End()
```
Comment (MrAlias) [inline sdk/trace/simple_span_processor_test.go:168]: nit: not sure this needs to be defered

```suggestion
		span.End()
```

File: sdk/trace/simple_span_processor_test.go [hunk 9]
```diff
@@ -150,6 +153,30 @@ func TestSimpleSpanProcessorShutdownOnEndConcurrency(t *testing.T) {
...
+	}
+
+	go span("test-span-1")
+	go span("test-span-2")
```
[the remark this answers is not in the document]
Comment (pellared) [reply sdk/trace/simple_span_processor_test.go:172]: What is the value of this?

**Review by hanyuancheung — APPROVED** [review]

**Comment by Kaushal28** [comment]
Hey @MrAlias, incorporated your suggestions. Could you please have a look?

File: sdk/trace/simple_span_processor_test.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (pellared) [reply sdk/trace/simple_span_processor_test.go]: @MrAlias What is the value of this?

[the remark this answers is not in the document]
Comment (MrAlias) [reply sdk/trace/simple_span_processor_test.go]: To catch a race panic in the `span` function.

[the remark this answers is not in the document]
Comment (MrAlias) [reply sdk/trace/simple_span_processor_test.go]: Yeah, that ^ looks right

**Comment by Kaushal28** [comment]
@pellared @MrAlias Thanks for pointing it out. I have fixed the test case as suggested. Please have a look. Thanks.

**Review by pellared — APPROVED** [review]

**Review by dmathieu — APPROVED** [review]

**Pull request merged**